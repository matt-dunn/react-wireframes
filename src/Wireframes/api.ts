/** !
 * Copyright (c) 2019, Matt Dunn
 *
 * @author Matt Dunn
 */

import { ComponentType, ReactNode } from "react";

import { getWireframeAnnotation, updateWireframeAnnotation } from "./utils";

type APIOptions = {
  updater: (annotations: WireframeAnnotations) => void;
  highlightNote?: (annotation?: WireframeAnnotation) => void;
}

export type WireframeAnnotationOptions = {
  title: ReactNode;
  description: ReactNode;
}

export type WireframeAnnotation = {
  id: number;
  Component: ComponentType<any>;
  count: number;
  options: WireframeAnnotationOptions;
}

export type WireframeAnnotations = WireframeAnnotation[];

type OpenCallback = (isOpen: boolean) => void;

export type WireframeAnnotationAPI = {
  setOptions: (options: APIOptions) => APIOptions;
  getAnnotations: () => WireframeAnnotations;
  register: (Component: ComponentType<any>, options: WireframeAnnotationOptions) => WireframeAnnotation;
  unregister: (Component: ComponentType<any>) => void;
  highlightNote: (Component: ComponentType<any> | undefined) => void;
  setOpen: (isOpen: boolean) => boolean;
  onOpen: (cb: OpenCallback) => {unregister: () => void};
  isOpen: () => boolean;
  setParentReference: (id?: number) => void;
  getParentReference: () => number | undefined;
}

export type withWireframeAnnotationProps = {
  annotationId?: number;
}

export function API(defaultOptions?: APIOptions): WireframeAnnotationAPI {
  let annotations: WireframeAnnotations = [];
  let apiOptions: APIOptions = defaultOptions || {} as APIOptions;
  let parentReference: number | undefined;

  let openCallbacks: OpenCallback[] = [];
  let isOpen = false;

  return {
    setOptions: (options) => {
      apiOptions = options;
      return apiOptions;
    },
    getAnnotations: () => annotations,
    register: (Component, options) => {
      const annotation = getWireframeAnnotation(annotations, Component);

      if (annotation) {
        const updatedAnnotation = {
          ...annotation,
          count: annotation.count + 1,
        };

        annotations = updateWireframeAnnotation(annotations, annotation, updatedAnnotation);

        /* istanbul ignore else */
        if (apiOptions.updater) {
          apiOptions.updater(annotations);
        }

        return updatedAnnotation;
      }

      const newAnnotation = {
        id: annotations.length + 1,
        Component,
        count: 1,
        options,
      };

      annotations = [...annotations, newAnnotation];

      if (apiOptions.updater) {
        apiOptions.updater(annotations);
      }

      return newAnnotation;
    },
    unregister: (Component) => {
      const annotation = getWireframeAnnotation(annotations, Component);

      if (annotation) {
        if (annotation.count > 1) {
          annotations = updateWireframeAnnotation(annotations, annotation, {
            ...annotation,
            count: annotation.count - 1,
          });
        } else {
          annotations = annotations.filter(c => c !== annotation);
        }

        if (apiOptions.updater) {
          apiOptions.updater(annotations);
        }
      }
    },
    highlightNote: Component => apiOptions && apiOptions.highlightNote && apiOptions.highlightNote(getWireframeAnnotation(annotations, Component)),
    setOpen: (open) => {
      if (open !== isOpen) {
        isOpen = open;

        openCallbacks.forEach(cb => cb(isOpen));
      }

      return isOpen;
    },
    onOpen: (cb) => {
      openCallbacks = [...openCallbacks, cb];

      return {
        unregister: () => {
          openCallbacks = openCallbacks.filter(c => c !== cb);
        },
      };
    },
    isOpen: () => isOpen,
    setParentReference: (id) => {
      parentReference = id;
    },
    getParentReference: () => parentReference,
  };
}
