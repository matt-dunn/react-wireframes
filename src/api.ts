/** !
 * Copyright (c) 2019, Matt Dunn
 *
 * @author Matt Dunn
 */

import { ComponentType, ReactNode } from "react";
import { remove } from "lodash";

type APIOptions = {
    updater: (components: WireFrameComponents) => void;
    highlightNote?: (component?: WireFrameComponent) => void;
}

export type WireFrameComponentOptions = {
    title: ReactNode;
    description: ReactNode;
}

export type WireFrameComponent = {
    id: number;
    Component: ComponentType<any>;
    count: number;
    options: WireFrameComponentOptions;
}

export type WireFrameComponents = WireFrameComponent[];

type OpenCallback = (isOpen: boolean) => void;

export type WireFrameAnnotationAPI = {
    setOptions: (options: APIOptions) => APIOptions;
    getComponents: () => WireFrameComponents;
    register: (Component: ComponentType<any>, options: WireFrameComponentOptions) => WireFrameComponent;
    unregister: (Component: ComponentType<any>) => void;
    highlightNote: (Component: ComponentType<any> | undefined) => void;
    setOpen: (isOpen: boolean) => boolean;
    onOpen: (cb: OpenCallback) => {unregister: () => void};
    isOpen: () => boolean;
}

export const API = function API(defaultOptions?: APIOptions): WireFrameAnnotationAPI {
  let components: WireFrameComponents = [];
  let apiOptions: APIOptions = defaultOptions || {} as APIOptions;

  let openCallbacks: OpenCallback[] = [];
  let isOpen = false;

  const getComponent = (Component: ComponentType<any> | undefined): WireFrameComponent | undefined => Component && components.find(c => c.Component === Component);

  return {
    setOptions: (options) => {
      apiOptions = options;
      return apiOptions;
    },
    getComponents: () => components,
    register: (Component, options) => {
      const component = getComponent(Component);

      if (component) {
        component.count += 1;

        if (apiOptions.updater) {
          apiOptions.updater(components);
        }

        return component;
      }

      const newComponent = {
        id: components.length + 1,
        Component,
        count: 1,
        options,
      };

      components = [...components, newComponent];

      if (apiOptions.updater) {
        apiOptions.updater(components);
      }

      return newComponent;
    },
    unregister: (Component) => {
      const component = getComponent(Component);

      if (component) {
        component.count -= 1;

        if (component.count === 0) {
          components = remove(components, c => c.Component !== Component);
        }
      }

      if (apiOptions.updater) {
        apiOptions.updater(components);
      }
    },
    highlightNote: Component => apiOptions && apiOptions.highlightNote && apiOptions.highlightNote(getComponent(Component)),
    setOpen: (open) => {
      isOpen = open;
      openCallbacks.forEach(cb => cb(isOpen));
      return isOpen;
    },
    onOpen: (cb) => {
      openCallbacks = [...openCallbacks, cb];
      return {
        unregister: () => {
          openCallbacks = remove(openCallbacks, c => c !== cb);
        },
      };
    },
    isOpen: () => isOpen,
  };
};
