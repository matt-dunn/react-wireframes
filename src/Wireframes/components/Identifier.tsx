/** !
 * Copyright (c) 2019, Matt Dunn
 *
 * @author Matt Dunn
 */

import React from "react";
import styled from "@emotion/styled";

import { WireframeAnnotation } from "../api";

type IdentifierProps = {
  annotation: WireframeAnnotation;
  className?: string;
}

export const IdentifierContainer = styled.cite`
  all: initial;
  box-sizing: border-box;
  border-radius: 2em;
  color: #333;
  background-color: yellow;
  border-color: #caca00;
  white-space: nowrap;
  min-width: 2em;
  max-width: initial;
  min-height: auto;
  max-height: initial;
  height: 2em;
  padding: 0.25em 0.5em;
  margin: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  position: initial;
  overflow: visible;
  font-style: normal;
  font-weight: normal;
  font-size: 1rem;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",Arial,"Noto Sans",sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji";
  text-align: center;
  text-decoration: none;
  text-transform: none;
  flex-grow: 0;
  flex-shrink: 0;
`;

export const Identifier = ({ annotation, className }: IdentifierProps) => (
  <IdentifierContainer data-annotation-identifier className={className}>{annotation.id.toLocaleString()}</IdentifierContainer>
);
