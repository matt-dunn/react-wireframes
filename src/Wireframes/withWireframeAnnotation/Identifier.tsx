/** !
 * Copyright (c) 2019, Matt Dunn
 *
 * @author Matt Dunn
 */

import React from "react";
import styled from "@emotion/styled";
import { CSSTransition } from "react-transition-group";

import Identifier from "../components/Identifier";
import { WireframeAnnotation } from "../api";

type IdentifierProps = {
  annotation: WireframeAnnotation;
  show: boolean;
}

const IdentifierContainer = styled(Identifier)`
  position: absolute;
  top: -1em;
  left: -1em;
  z-index: 4000;
  cursor: default;
  transition: opacity 250ms, visibility 250ms;
  opacity: 0.75;
  visibility: visible;

  &.fade-Identifier.enter {
    opacity: 0;
    visibility: hidden;
  }
  &.fade-Identifier.enter-active {
    opacity: 0.75;
    visibility: visible;
  }
  &.fade-Identifier.exit {
    opacity: 0.75;
    visibility: visible;
  }
  &.fade-Identifier.exit-active {
    opacity: 0;
    visibility: hidden;
  }
`;

export const IdentifierComponent = ({ annotation, show = true }: IdentifierProps) => (
  <CSSTransition
    timeout={250}
    className="fade-Identifier"
    in={show}
    mountOnEnter
    unmountOnExit
  >
    <IdentifierContainer data-annotation-identifier annotation={annotation} />
  </CSSTransition>
);

export default IdentifierComponent;
