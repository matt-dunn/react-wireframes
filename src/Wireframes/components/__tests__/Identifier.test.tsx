/** !
 * Copyright (c) 2019, Matt Dunn
 *
 * @author Matt Dunn
 */

import React from "react";
import { shallow } from "enzyme";

import { WireframeAnnotation, WireframeAnnotationAPI } from "../../api";

import { Identifier as Component } from "../Identifier";

describe("Wireframe: WireframeAnnotationsNote", () => {
  let annotation: WireframeAnnotation;
  let MockedComponent;

  beforeEach(() => {
    MockedComponent = jest.fn();

    annotation = {
      id: 1,
      count: 1,
      Component: MockedComponent,
      options: {
        title: "Test component",
        description: "Test description",
      },
    };
  });

  it("should render correctly", () => {
    const wrapper = shallow(
      <Component
        annotation={annotation}
      />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("should render ID with parent ID", () => {
    const wrapper = shallow(
      <Component
        annotation={{
          ...annotation,
          id: 12,
        }}
        parentReference={{
          id: 1,
          api: {
            getParentReference: () => {},
          } as WireframeAnnotationAPI,
        }}
      />,
    );

    expect(wrapper.text()).toBe("1.12");
  });

  it("should render ID with nested parent ID", () => {
    const wrapper = shallow(
      <Component
        annotation={{
          ...annotation,
          id: 12,
        }}
        parentReference={{
          id: 1,
          api: {
            getParentReference: () => ({
              id: 2,
              api: {
                getParentReference: () => {},
              } as WireframeAnnotationAPI,
            }),
          } as WireframeAnnotationAPI,
        }}
      />,
    );

    expect(wrapper.text()).toBe("2.1.12");
  });
});
