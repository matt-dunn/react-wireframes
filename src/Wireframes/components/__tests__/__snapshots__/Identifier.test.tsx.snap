// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wireframe: WireframeAnnotationsNote should render correctly 1`] = `
<IdentifierContainer
  data-annotation-identifier={true}
>
  1
</IdentifierContainer>
`;
